compiler/src/dotty/tools/dotc/core/Constants.scala

```diff
@@ -161,25 +163,29 @@ object Constants {
           }
         case pt => pt
       }
-      val target = classBound(pt).typeSymbol
-      if (target == tpe.typeSymbol)
-        this
-      else if ((target == defn.ByteClass) && isByteRange)
-        Constant(byteValue)
-      else if (target == defn.ShortClass && isShortRange)
-        Constant(shortValue)
-      else if (target == defn.CharClass && isCharRange)
-        Constant(charValue)
-      else if (target == defn.IntClass && isIntRange)
-        Constant(intValue)
-      else if (target == defn.LongClass && isLongRange)
-        Constant(longValue)
-      else if (target == defn.FloatClass && isFloatRange)
-        Constant(floatValue)
-      else if (target == defn.DoubleClass && isNumeric)
-        Constant(doubleValue)
-      else
-        null
+      pt match
+        case ConstantType(value) if value == this => this
+        case _: SingletonType => null
+        case _ =>
+          val target = classBound(pt).typeSymbol
+          if (target == tpe.typeSymbol)
+            this
+          else if ((target == defn.ByteClass) && isByteRange)
+            Constant(byteValue)
+          else if (target == defn.ShortClass && isShortRange)
+            Constant(shortValue)
+          else if (target == defn.CharClass && isCharRange)
+            Constant(charValue)
+          else if (target == defn.IntClass && isIntRange)
+            Constant(intValue)
+          else if (target == defn.LongClass && isLongRange)
+            Constant(longValue)
+          else if (target == defn.FloatClass && isFloatRange)
+            Constant(floatValue)
+          else if (target == defn.DoubleClass && isNumeric)
+            Constant(doubleValue)
```

Comment:
In a series of `if p && q then` where the ps are mutually exclusive, preferable to
```
expr match p => if q then res else default
```
Noticed the similar pattern in constant folding. Maybe nbd for 7 tests. My question would be what should my fingers type automatically for this pattern.

It could optimize `eq` as exclusive but maybe not `==`.